lib/improver/convection.py

```diff
@@ -0,0 +1,322 @@
+                calculated.
+            threshold : float
+                The threshold that will be applied.
+
+        Returns:
+            cubelist : Iris.cube.CubeList
+                Cubelist containing cubes with the absolute difference
+                between adjacent grid squares along x and y, respectively.
+        """
+        diff_along_x_cube, diff_along_y_cube = (
+            DifferenceBetweenAdjacentGridSquares().process(cube))
+        # Compute the absolute values of the differences to ensure that
+        # negative differences are included.
+        diff_along_x_cube.data = np.absolute(diff_along_x_cube.data)
+        diff_along_y_cube.data = np.absolute(diff_along_y_cube.data)
+        cubelist = iris.cube.CubeList([diff_along_x_cube, diff_along_y_cube])
+        return cubelist
+
+    def iterate_over_threshold(self, cubelist, threshold):
+        """
+        Iterate over the application of thresholding to multiple cubes.
+
+        Args:
+            cubelist : Iris.cube.CubeList
+                Cubelist containing cubes to be thresholded.
+            threshold : float
+                The threshold that will be applied.
+
+        Returns:
+            cubes : Iris.cube.CubeList
+                Cubelist after thresholding each cube.
+        """
+        cubes = iris.cube.CubeList([])
+        for cube in cubelist:
+            threshold_cube = (
+                BasicThreshold(
+                    threshold, fuzzy_factor=self.fuzzy_factor,
+                    below_thresh_ok=self.below_thresh_ok
+                    ).process(cube.copy()))
+            cubes.append(threshold_cube)
+        return cubes
+
+    @staticmethod
+    def sum_differences_between_adjacent_grid_squares(cube, thresholded_cubes):
+        """
+        Put the differences back onto the original grid by summing together
+        the array with offsets. This covers the fact that the difference
+        cubes will result in output on a staggered grid compared with the
+        input cube.
+
+        Args:
+            cube : Iris.cube.Cube
+                The cube with the original grid.
+            thresholded_cubes : Iris.cube.CubeList
+                Cubelist containing differences between adjacent grid squares
+                along x and differences between adjacent grid squares along y,
+                which have been thresholded.
+
+        Args:
+            cube_on_orig_grid : Iris.cube.Cube
+                Cube on the original grid with the values from the thresholded
+                adjacent grid square difference cubes inserted. The resulting
+                values have been restricted to be between 0 and 1.
+        """
+        threshold_cube_x, threshold_cube_y = thresholded_cubes
+        cube_on_orig_grid = cube.copy()
+        cube_on_orig_grid.data = np.zeros(cube_on_orig_grid.shape)
+        cube_on_orig_grid.data[..., :-1, :] += threshold_cube_y.data
```

Comment:
The edge rows do get filled, if there are non-zero values within `threshold_cube_y`. As `threshold_cube_y` is only one row shorter than `cube_on_orig_grid`, the values at (0,0) from `threshold_cube_y` are put into both (0,0) and (1,0) within `cube_on_orig_grid`, so the edge rows should still always to filled.